feat(api): Allow health endpoint to return 503s when Vector is shutting down

spencergilbert opened the pull request:
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

Adds an Arc wrapped AtomicBool for the `stop` function from `RunningTopology` to use. The health handler references this new field to determine if a response is 200 or 503.

Initial work from health check RFC before larger task described in https://github.com/vectordotdev/vector/issues/10555

<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/timberio/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

jszwedko reviewed (commented):
Nice! Implementation seems sensible to me though others will have more informed opinions. Would it be feasible to add a test for this though? Maybe in https://github.com/vectordotdev/vector/blob/master/tests/shutdown.rs ?

leebenson commented:
While this approach works, it feels a little redundant to have an atomic bool who's purpose is to communicate what a `SignalTo` already gives us.

IMO, it'd be cleaner to switch the signal channel to a MPMC (i.e. [broadcast](https://docs.rs/tokio/latest/tokio/sync/broadcast/index.html)), pass the server a `SignalRx`, and then match the `.try_recv` in the health handler to determine if `SignalTo::Shutdown | SignalTo::Quit` has been passed.

If we wanted to expand health to accommodate reload information, that would also be present on the `SignalTo` enum. We could then represent status more accurately inside the GraphQL API, too.

blt commented on src/api/handler.rs:
```diff
@@ -1,7 +1,19 @@
-pub async fn health() -> Result<impl Reply, Rejection> {
-    Ok(json(&json!({"ok": true})))
+pub async fn health(running: Arc<AtomicBool>) -> Result<impl Reply, Rejection> {
+    if running.load(std::sync::atomic::Ordering::Relaxed) {
```
Maybe go ahead and import up to `atomic`?

spencergilbert replied:
I thought I had, might have missed it there 👍

blt reviewed (commented):
This works for me, modulo the cilppy dings. The implementation is simple and direct. I could see a use for a more elaborate communication mechanism -- a broadcast, like Lee suggested -- but there doesn't seem to be motivation for it yet, unless there's requirements I'm unaware of. If we had more elaborate goals around shutdown, or more explicit guarantees for users about how it worked, then I could see the complexity of this needing to ramp. As is, I'm inclined to +1 this and suggest we make an RFC for how shutdown needs to work if this is too simplistic to support the use cases we intend, since that'll have impact on how we fiddle with the topology in core and we need to coordinate.

spencergilbert commented:
@blt, fixed the clippy errors and cleaned up the `use` statements. @leebenson I'm also of the opinion that we could use some more discussion into the per component setup as based on what I was seeing while working on this I wouldn't necessarily attach that info to the SignalTo enum.

@jszwedko did you want something like a unit test or an integration test?

jszwedko commented:
> @jszwedko did you want something like a unit test or an integration test?

I think an integration test like the ones you see in `test/shutdown.rs` would be appropriate: just asserting that the API returns a 503 when Vector is shutting down.

leebenson commented:
> I'm also of the opinion that we could use some more discussion into the per component setup as based on what I was seeing while working on this I wouldn't necessarily attach that info to the SignalTo enum.

That sounds reasonable 👍🏻 

I don't think `SignalTo` is the right place to attach extended health information. It was mostly just leveraging a channel we already have. A MPMC strategy may be something to bear in mind when we have more concrete requirements.

jszwedko reviewed (approved):
Nice work! The test looks good.

leebenson commented on tests/shutdown.rs:
```diff
@@ -579,3 +580,55 @@ fn timely_reload_shutdown() {
+        .unwrap();
+
+    // Give vector time to start.
+    sleep(STARTUP_TIME);
```
I know this was here already, but this feels like a potential source for flaky tests. We could use the health endpoint to check for `{ok: true}` on a loop with a timeout, and maybe shave some secs off each integration test.

leebenson reviewed (approved):
Looks great, nice work! 🎉
